app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/alwayson/RealAlwaysOnStatusMonitor.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.mobile.android.vpn.alwayson
+
+import com.duckduckgo.di.scopes.ActivityScope
+import com.duckduckgo.mobile.android.vpn.VpnFeaturesRegistry
+import com.duckduckgo.mobile.android.vpn.dao.VpnHeartBeatDao
+import com.duckduckgo.mobile.android.vpn.dao.VpnServiceStateStatsDao
+import com.duckduckgo.mobile.android.vpn.heartbeat.VpnServiceHeartbeatMonitor
+import com.duckduckgo.mobile.android.vpn.model.VpnServiceState
+import com.duckduckgo.mobile.android.vpn.model.VpnStoppingReason
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+@ContributesBinding(ActivityScope::class)
+class RealAlwaysOnStatusMonitor @Inject constructor(
+    private val vpnHeartBeatDao: VpnHeartBeatDao,
+    private val vpnServiceStateDao: VpnServiceStateStatsDao,
+    private val vpnFeaturesRegistry: VpnFeaturesRegistry,
+) : AlwaysOnStatusMonitor {
+    override suspend fun isAlwaysOnEnabled(): Boolean {
+        return vpnServiceStateDao.getLastStateStats()?.alwaysOnState?.alwaysOnEnabled ?: false
+    }
+
+    override suspend fun vpnLastDisabledByAndroid(): Boolean {
+        fun vpnUnexpectedlyDisabled(): Boolean {
+            return vpnServiceStateDao.getLastStateStats()?.let {
+                (
+                    it.state == VpnServiceState.DISABLED &&
+                        it.stopReason != VpnStoppingReason.SELF_STOP &&
+                        it.stopReason != VpnStoppingReason.REVOKED
+                    )
+            } ?: false
+        }
+
+        fun vpnKilledBySystem(): Boolean {
+            val lastHeartBeat = vpnHeartBeatDao.hearBeats().maxByOrNull { it.timestamp }
+            return lastHeartBeat?.type == VpnServiceHeartbeatMonitor.DATA_HEART_BEAT_TYPE_ALIVE &&
+                !vpnFeaturesRegistry.isAnyFeatureRegistered()
```

Comment:
This used to be `!vpnFeaturesRegistry.isFeatureRegistered(AppTpVpnFeature.APPTP_VPN)`. I think this check will scale better.